File: src/main/kotlin/me/hippodev/tracking/StatsLogger.kt
package me.hippodev.tracking

import me.hippodev.config.StatsLoggingConfig
import me.hippodev.routing.MetricsSnapshot
import org.slf4j.LoggerFactory
import java.io.File
import java.sql.Connection
import java.sql.DriverManager
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * Optional, off-by-default local time-series log of MCGate's live stats - a periodic
 * [MetricsSnapshot] (players online, per-host player counts, per-backend active connections/
 * latency, connection-tracking queue/drop counters) written to a local SQLite database with a
 * timestamp, distinct from [ConnectionTracker]'s per-session history. Lets traffic trends be
 * queried locally (e.g. "players online per host, over the last week") without standing up a
 * separate Prometheus server against the `/metrics` API endpoint.
 *
 * Unlike [ConnectionTracker] (which can see a burst of hundreds of records at once, from many
 * Netty threads, and needs a bounded queue + batched writer to survive that), this only ever
 * writes one snapshot every [StatsLoggingConfig.intervalMillis] (60s by default) from a single
 * dedicated scheduled-executor thread - there's no concurrent-producer backpressure problem to
 * solve here, so a plain scheduled task doing a direct write is enough. It still can't ever touch
 * a Netty event-loop thread: gathering a snapshot ([me.hippodev.routing.collectMetrics]) and
 * writing it both happen entirely on this dedicated thread.
 */
object StatsLogger {
    private val log = LoggerFactory.getLogger(StatsLogger::class.java)

    @Volatile private var config: StatsLoggingConfig = StatsLoggingConfig()
    @Volatile private var connection: Connection? = null
    private var scheduler: ScheduledExecutorService? = null

    val enabled: Boolean get() = connection != null

    /** Applies (or re-applies) [newConfig], starting/stopping/restarting the scheduled snapshot
     *  task as needed - same idempotent pattern as [ConnectionTracker.applyConfig], safe to call
     *  on every config (re)load. [snapshotSupplier] is called fresh on every tick, not cached, so
     *  it always reflects whatever routes/runtime are current at that moment even across a config
     *  reload that swaps them out. */
    @Synchronized
    fun applyConfig(newConfig: StatsLoggingConfig, snapshotSupplier: () -> MetricsSnapshot) {
        if (newConfig == config) return
        stopInternal()
        config = newConfig
        if (newConfig.enabled) startInternal(newConfig, snapshotSupplier)
    }

    @Synchronized
    fun shutdown() {
        stopInternal()
    }

    private fun startInternal(cfg: StatsLoggingConfig, snapshotSupplier: () -> MetricsSnapshot) {
        try {
            val file = File(cfg.dbPath)
            file.absoluteFile.parentFile?.mkdirs()
            val conn = DriverManager.getConnection("jdbc:sqlite:${file.path}")
            conn.createStatement().use { st ->
                st.execute("PRAGMA journal_mode=WAL")
                st.execute("PRAGMA synchronous=NORMAL")
                st.execute("PRAGMA auto_vacuum=INCREMENTAL")
                st.execute("PRAGMA foreign_keys=ON")
                st.execute(
                    """
                    CREATE TABLE IF NOT EXISTS stats_snapshots (
                        id INTEGER PRIMARY KEY AUTOINCREMENT,
                        timestamp INTEGER NOT NULL,
                        players_online INTEGER NOT NULL,
                        uptime_seconds REAL NOT NULL,
                        tracking_enabled INTEGER NOT NULL,
                        tracking_queued_records INTEGER NOT NULL,
                        tracking_dropped_records_total INTEGER NOT NULL
                    )
                    """.trimIndent()
                )
                st.execute("CREATE INDEX IF NOT EXISTS idx_stats_snapshots_timestamp ON stats_snapshots(timestamp)")
                st.execute(
                    """
                    CREATE TABLE IF NOT EXISTS stats_host_snapshots (
                        id INTEGER PRIMARY KEY AUTOINCREMENT,
                        snapshot_id INTEGER NOT NULL REFERENCES stats_snapshots(id) ON DELETE CASCADE,
                        host TEXT NOT NULL,
                        players_online INTEGER NOT NULL
                    )
                    """.trimIndent()
                )
                st.execute("CREATE INDEX IF NOT EXISTS idx_stats_host_snapshots_snapshot_id ON stats_host_snapshots(snapshot_id)")
                st.execute(
                    """
                    CREATE TABLE IF NOT EXISTS stats_backend_snapshots (
                        id INTEGER PRIMARY KEY AUTOINCREMENT,
                        snapshot_id INTEGER NOT NULL REFERENCES stats_snapshots(id) ON DELETE CASCADE,
                        route_index INTEGER NOT NULL,
                        hosts TEXT NOT NULL,
                        backend TEXT NOT NULL,
                        active_connections INTEGER NOT NULL,
                        latency_ms INTEGER
                    )
                    """.trimIndent()
                )
                st.execute("CREATE INDEX IF NOT EXISTS idx_stats_backend_snapshots_snapshot_id ON stats_backend_snapshots(snapshot_id)")
            }
            connection = conn

            val exec = Executors.newSingleThreadScheduledExecutor { r ->
                Thread(r, "stats-logger").apply { isDaemon = true }
            }
            // scheduleWithFixedDelay, not AtFixedRate: if a single write ever takes longer than
            // intervalMillis (a slow disk), this waits for it to finish rather than piling up
            // back-to-back ticks queued on the same single thread. Pruning runs every tick too,
            // right after the write - both are cheap (one snapshot's worth of rows; two DELETEs
            // and an incremental vacuum), and snapshots are already infrequent by design, so
            // there's no need for a separate, more complex "prune roughly hourly" schedule.
            exec.scheduleWithFixedDelay(
                {
                    runCatching { writeSnapshot(snapshotSupplier()) }.onFailure { log.warn("Stats logging write failed", it) }
                    runCatching { prune(cfg) }.onFailure { log.warn("Stats logging prune failed", it) }
                },
                cfg.intervalMillis, cfg.intervalMillis, TimeUnit.MILLISECONDS
            )
            scheduler = exec

            log.info("Stats logging enabled, writing to {} every {}ms", file.path, cfg.intervalMillis)
        } catch (e: Exception) {
            log.error("Failed to start stats logging, feature stays disabled", e)
            connection?.let { runCatching { it.close() } }
            connection = null
        }
    }

    private fun stopInternal() {
        scheduler?.let { exec ->
            exec.shutdown()
            // Bounded wait, same reasoning as ConnectionTracker.stopInternal - a graceful
            // shutdown must still terminate in finite time even if a write is stuck.
            runCatching { exec.awaitTermination(5, TimeUnit.SECONDS) }
        }
        scheduler = null
        connection?.let { runCatching { it.close() } }
        connection = null
    }

    private fun writeSnapshot(s: MetricsSnapshot) {
        val conn = connection ?: return
        conn.autoCommit = false
        try {
            val snapshotId = conn.prepareStatement(
                """INSERT INTO stats_snapshots
                   (timestamp, players_online, uptime_seconds, tracking_enabled, tracking_queued_records, tracking_dropped_records_total)
                   VALUES (?,?,?,?,?,?)""",
                java.sql.Statement.RETURN_GENERATED_KEYS
            ).use { ps ->
                ps.setLong(1, s.timestamp)
                ps.setInt(2, s.playersOnline)
                ps.setDouble(3, s.uptimeSeconds)
                ps.setInt(4, if (s.trackingEnabled) 1 else 0)
                ps.setInt(5, s.trackingQueuedRecords)
                ps.setLong(6, s.trackingDroppedRecordsTotal)
                ps.executeUpdate()
                ps.generatedKeys.use { rs -> rs.next(); rs.getLong(1) }
            }

            if (s.onlineByHost.isNotEmpty()) {
                conn.prepareStatement("INSERT INTO stats_host_snapshots (snapshot_id, host, players_online) VALUES (?,?,?)").use { ps ->
                    for ((host, count) in s.onlineByHost) {
                        ps.setLong(1, snapshotId)
                        ps.setString(2, host)
                        ps.setInt(3, count)
                        ps.addBatch()
                    }
                    ps.executeBatch()
                }
            }

            if (s.backends.isNotEmpty()) {
                conn.prepareStatement(
                    """INSERT INTO stats_backend_snapshots
                       (snapshot_id, route_index, hosts, backend, active_connections, latency_ms) VALUES (?,?,?,?,?,?)"""
                ).use { ps ->
                    for (b in s.backends) {
                        ps.setLong(1, snapshotId)
                        ps.setInt(2, b.routeIndex)
                        ps.setString(3, b.hosts.joinToString(","))
                        ps.setString(4, b.backend)
                        ps.setInt(5, b.active)
                        if (b.latencyMillis != null) ps.setLong(6, b.latencyMillis) else ps.setNull(6, java.sql.Types.INTEGER)
                        ps.addBatch()
                    }
                    ps.executeBatch()
                }
            }

            conn.commit()
        } catch (e: Exception) {
            runCatching { conn.rollback() }
            throw e
        } finally {
            conn.autoCommit = true
        }
    }

    private fun prune(cfg: StatsLoggingConfig) {
        val conn = connection ?: return
        val cutoff = System.currentTimeMillis() - cfg.retentionDays * 86_400_000L
        conn.prepareStatement("DELETE FROM stats_snapshots WHERE timestamp < ?").use { ps ->
            ps.setLong(1, cutoff)
            ps.executeUpdate()
        }
        conn.prepareStatement(
            "DELETE FROM stats_snapshots WHERE id IN (SELECT id FROM stats_snapshots ORDER BY id DESC LIMIT -1 OFFSET ?)"
        ).use { ps ->
            ps.setInt(1, cfg.maxRecords)
            ps.executeUpdate()
        }
        // stats_host_snapshots/stats_backend_snapshots rows are cleaned up automatically by the
        // ON DELETE CASCADE foreign keys above (PRAGMA foreign_keys=ON, set at connection open).
        conn.createStatement().use { st -> st.execute("PRAGMA incremental_vacuum(2000)") }
    }
}

File: src/state/stats_logger.rs
//! Port of `tracking/StatsLogger.kt` — optional, off-by-default local time-series log of
//! MCGate's live stats, written to SQLite on a fixed interval.
//!
//! Unlike `connection_tracker.rs` (which can see a burst of records at once, from many
//! concurrent tasks, and needs a bounded queue + batched writer to survive that), this only ever
//! writes one snapshot every `interval_millis` from a single dedicated thread — no
//! concurrent-producer backpressure problem to solve, so a plain timed loop doing a direct write
//! is enough.

use std::sync::{Arc, Mutex, OnceLock};
use std::time::Duration;

use rusqlite::Connection;

use crate::state::app_state::AppState;
use crate::config::StatsLoggingConfig;
use crate::state::MetricsSnapshot;

struct Inner {
    shutdown: crossbeam_channel::Sender<()>,
    handle: std::thread::JoinHandle<()>,
}

#[derive(Default)]
pub struct StatsLogger {
    inner: Mutex<Option<Inner>>,
    current_config: Mutex<StatsLoggingConfig>,
}

pub fn stats_logger() -> &'static StatsLogger {
    static INSTANCE: OnceLock<StatsLogger> = OnceLock::new();
    INSTANCE.get_or_init(StatsLogger::default)
}

impl StatsLogger {
    pub fn is_enabled(&self) -> bool {
        self.inner.lock().unwrap().is_some()
    }

    /// Applies (or re-applies) `new_config`, starting/stopping/restarting the scheduled snapshot
    /// task as needed — same idempotent pattern as `ConnectionTracker::apply_config`. `state`
    /// supplies both the routes/runtime `collect_metrics` needs and the tokio `Handle` the
    /// (plain `std::thread`) writer bridges into it with.
    pub fn apply_config(&'static self, new_config: StatsLoggingConfig, state: Arc<AppState>) {
        {
            let current = self.current_config.lock().unwrap();
            if *current == new_config {
                return;
            }
        }
        self.stop_internal();
        *self.current_config.lock().unwrap() = new_config.clone();
        if new_config.enabled {
            self.start_internal(new_config, state);
        }
    }

    pub fn shutdown(&self) {
        self.stop_internal();
    }

    fn start_internal(&'static self, cfg: StatsLoggingConfig, state: Arc<AppState>) {
        let (shutdown_tx, shutdown_rx) = crossbeam_channel::bounded::<()>(1);
        let db_path = cfg.db_path.clone();
        let handle = match open_connection(&db_path) {
            Ok(conn) => {
                let handle = std::thread::Builder::new()
                    .name("stats-logger".into())
                    .spawn(move || writer_loop(conn, shutdown_rx, cfg, state))
                    .expect("failed to spawn stats-logger thread");
                tracing::info!("Stats logging enabled, writing to {db_path}");
                handle
            }
            Err(e) => {
                tracing::error!("Failed to start stats logging, feature stays disabled: {e}");
                return;
            }
        };
        *self.inner.lock().unwrap() = Some(Inner { shutdown: shutdown_tx, handle });
    }

    fn stop_internal(&self) {
        let inner = self.inner.lock().unwrap().take();
        if let Some(inner) = inner {
            let _ = inner.shutdown.send(());
            let _ = inner.handle.join();
        }
    }
}

fn open_connection(db_path: &str) -> rusqlite::Result<Connection> {
    if let Some(parent) = std::path::Path::new(db_path).parent() {
        if !parent.as_os_str().is_empty() {
            let _ = std::fs::create_dir_all(parent);
        }
    }
    let conn = Connection::open(db_path)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL;
         PRAGMA synchronous=NORMAL;
         PRAGMA auto_vacuum=INCREMENTAL;
         PRAGMA foreign_keys=ON;
         CREATE TABLE IF NOT EXISTS stats_snapshots (
             id INTEGER PRIMARY KEY AUTOINCREMENT,
             timestamp INTEGER NOT NULL,
             players_online INTEGER NOT NULL,
             uptime_seconds REAL NOT NULL,
             tracking_enabled INTEGER NOT NULL,
             tracking_queued_records INTEGER NOT NULL,
             tracking_dropped_records_total INTEGER NOT NULL
         );
         CREATE INDEX IF NOT EXISTS idx_stats_snapshots_timestamp ON stats_snapshots(timestamp);
         CREATE TABLE IF NOT EXISTS stats_host_snapshots (
             id INTEGER PRIMARY KEY AUTOINCREMENT,
             snapshot_id INTEGER NOT NULL REFERENCES stats_snapshots(id) ON DELETE CASCADE,
             host TEXT NOT NULL,
             players_online INTEGER NOT NULL
         );
         CREATE INDEX IF NOT EXISTS idx_stats_host_snapshots_snapshot_id ON stats_host_snapshots(snapshot_id);
         CREATE TABLE IF NOT EXISTS stats_backend_snapshots (
             id INTEGER PRIMARY KEY AUTOINCREMENT,
             snapshot_id INTEGER NOT NULL REFERENCES stats_snapshots(id) ON DELETE CASCADE,
             route_index INTEGER NOT NULL,
             hosts TEXT NOT NULL,
             backend TEXT NOT NULL,
             active_connections INTEGER NOT NULL,
             latency_ms INTEGER
         );
         CREATE INDEX IF NOT EXISTS idx_stats_backend_snapshots_snapshot_id ON stats_backend_snapshots(snapshot_id);",
    )?;
    Ok(conn)
}

fn writer_loop(mut conn: Connection, shutdown_rx: crossbeam_channel::Receiver<()>, cfg: StatsLoggingConfig, state: Arc<AppState>) {
    let interval = Duration::from_millis(cfg.interval_millis.max(1) as u64);
    let handle = state.runtime_handle().clone();
    loop {
        crossbeam_channel::select! {
            recv(shutdown_rx) -> _ => break,
            default(interval) => {}
        }
        let snapshot = handle.block_on(collect_snapshot(&state));
        if let Err(e) = write_snapshot(&mut conn, &snapshot) {
            tracing::warn!("Stats logging write failed: {e}");
        }
        if let Err(e) = prune(&conn, &cfg) {
            tracing::warn!("Stats logging prune failed: {e}");
        }
    }
}

async fn collect_snapshot(state: &Arc<AppState>) -> MetricsSnapshot {
    let cfg = state.config();
    let state = state.clone();
    crate::state::collect_metrics(&cfg.routes, move |r| cfg_route_runtime(&state, r)).await
}

fn cfg_route_runtime(state: &Arc<AppState>, route: &crate::config::Route) -> Option<Arc<crate::state::RouteRuntime>> {
    let cfg = state.config();
    cfg.routes.iter().position(|r| r == route).and_then(|i| state.route_runtime(i))
}

fn write_snapshot(conn: &mut Connection, s: &MetricsSnapshot) -> rusqlite::Result<()> {
    let tx = conn.transaction()?;
    let snapshot_id: i64 = {
        tx.execute(
            "INSERT INTO stats_snapshots
             (timestamp, players_online, uptime_seconds, tracking_enabled, tracking_queued_records, tracking_dropped_records_total)
             VALUES (?1,?2,?3,?4,?5,?6)",
            rusqlite::params![s.timestamp_millis, s.players_online as i64, s.uptime_seconds, s.tracking_enabled as i32, s.tracking_queued_records, s.tracking_dropped_records_total],
        )?;
        tx.last_insert_rowid()
    };

    if !s.online_by_host.is_empty() {
        let mut stmt = tx.prepare("INSERT INTO stats_host_snapshots (snapshot_id, host, players_online) VALUES (?1,?2,?3)")?;
        for (host, count) in &s.online_by_host {
            stmt.execute(rusqlite::params![snapshot_id, host, *count as i64])?;
        }
    }

    if !s.backends.is_empty() {
        let mut stmt = tx.prepare(
            "INSERT INTO stats_backend_snapshots (snapshot_id, route_index, hosts, backend, active_connections, latency_ms) VALUES (?1,?2,?3,?4,?5,?6)",
        )?;
        for b in &s.backends {
            stmt.execute(rusqlite::params![snapshot_id, b.route_index as i64, b.hosts.join(","), b.backend, b.active, b.latency_millis])?;
        }
    }

    tx.commit()
}

fn prune(conn: &Connection, cfg: &StatsLoggingConfig) -> rusqlite::Result<()> {
    let cutoff = now_millis() - cfg.retention_days as i64 * 86_400_000;
    conn.execute("DELETE FROM stats_snapshots WHERE timestamp < ?1", rusqlite::params![cutoff])?;
    conn.execute(
        "DELETE FROM stats_snapshots WHERE id IN (SELECT id FROM stats_snapshots ORDER BY id DESC LIMIT -1 OFFSET ?1)",
        rusqlite::params![cfg.max_records],
    )?;
    // stats_host_snapshots/stats_backend_snapshots rows are cleaned up automatically by the
    // ON DELETE CASCADE foreign keys above (PRAGMA foreign_keys=ON, set at connection open).
    conn.execute_batch("PRAGMA incremental_vacuum(2000)")
}

fn now_millis() -> i64 {
    std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.as_millis() as i64).unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_snapshot() -> MetricsSnapshot {
        MetricsSnapshot {
            // A genuinely "now" timestamp - see connection_tracker.rs's identical fix for why a
            // small constant would already look ancient to retention pruning regardless of the
            // configured window.
            timestamp_millis: now_millis(),
            players_online: 3,
            online_by_host: std::collections::HashMap::from([("a.example.com".to_string(), 3usize)]),
            uptime_seconds: 12.5,
            backends: vec![],
            players: vec![],
            tracking_enabled: false,
            tracking_queued_records: 0,
            tracking_dropped_records_total: 0,
            route_metrics: vec![],
        }
    }

    #[test]
    fn schema_creation_and_write_round_trip() {
        let dir = std::env::temp_dir().join(format!("mcgate-statslogger-test-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let db_path = dir.join("stats.db");

        let mut conn = open_connection(db_path.to_str().unwrap()).unwrap();
        write_snapshot(&mut conn, &sample_snapshot()).unwrap();

        let count: i64 = conn.query_row("SELECT COUNT(*) FROM stats_snapshots", [], |r| r.get(0)).unwrap();
        assert_eq!(count, 1);
        let host_count: i64 = conn.query_row("SELECT COUNT(*) FROM stats_host_snapshots", [], |r| r.get(0)).unwrap();
        assert_eq!(host_count, 1);

        std::fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn prune_deletes_old_snapshots_and_cascades_host_rows() {
        let dir = std::env::temp_dir().join(format!("mcgate-statslogger-prune-test-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let db_path = dir.join("stats.db");

        let mut conn = open_connection(db_path.to_str().unwrap()).unwrap();
        let mut old = sample_snapshot();
        old.timestamp_millis = 0;
        write_snapshot(&mut conn, &old).unwrap();
        write_snapshot(&mut conn, &sample_snapshot()).unwrap();

        let cfg = StatsLoggingConfig { retention_days: 1, max_records: 200_000, ..StatsLoggingConfig::default() };
        prune(&conn, &cfg).unwrap();

        let count: i64 = conn.query_row("SELECT COUNT(*) FROM stats_snapshots", [], |r| r.get(0)).unwrap();
        assert_eq!(count, 1);
        let host_count: i64 = conn.query_row("SELECT COUNT(*) FROM stats_host_snapshots", [], |r| r.get(0)).unwrap();
        assert_eq!(host_count, 1, "cascade delete should have removed the pruned snapshot's host rows too");

        std::fs::remove_dir_all(&dir).unwrap();
    }
}
